third_party/infozip/infozip.mk

```diff
@@ -0,0 +1,134 @@
+#-*-mode:makefile-gmake;indent-tabs-mode:t;tab-width:8;coding:utf-8-*-┐
+#───vi: set et ft=make ts=8 tw=8 fenc=utf-8 :vi───────────────────────┘
+PKGS += THIRD_PARTY_ZIP
+
+THIRD_PARTY_ZIP_FILES = $(wildcard third_party/infozip/zip/*) third_party/infozip/zip/unix/unix.c
+THIRD_PARTY_ZIP_SRCS_C = $(filter %.c,$(THIRD_PARTY_ZIP_FILES))
+THIRD_PARTY_ZIP = $(THIRD_PARTY_ZIP_COMS) $(THIRD_PARTY_ZIP_COMS:%=%.dbg)
+
+THIRD_PARTY_ZIP_COMS = 							\
+	o/$(MODE)/third_party/infozip/zip.com 		\
+	o/$(MODE)/third_party/infozip/zipsplit.com 	\
+	o/$(MODE)/third_party/infozip/zipnote.com 	\
+	o/$(MODE)/third_party/infozip/zipcloak.com 	
+
+THIRD_PARTY_ZIP_OBJS = $(sort			\
+	$(THIRD_PARTY_ZIP_COM_OBJS)			\
+	$(THIRD_PARTY_ZIPCLOAK_OBJS)		\
+	$(THIRD_PARTY_ZIPNOTE_OBJS)			\
+	$(THIRD_PARTY_ZIPSPLIT_OBJS)		\
+	)
+
+THIRD_PARTY_ZIP_UTIL_OBJS1 = \
+	o/$(MODE)/third_party/infozip/zip/globals.o 		\
+	o/$(MODE)/third_party/infozip/zip/unix/unix_.o 		\
+	o/$(MODE)/third_party/infozip/zip/zipfile_.o 		\
+	o/$(MODE)/third_party/infozip/zip/fileio_.o 		\
+	o/$(MODE)/third_party/infozip/zip/util_.o 
+
+THIRD_PARTY_ZIP_UTIL_OBJS2 = \
+	o/$(MODE)/third_party/infozip/zip/crypt_.o 		\
+	o/$(MODE)/third_party/infozip/zip/crc32_.o
+
+THIRD_PARTY_ZIP_UTIL_OBJS = \
+	$(THIRD_PARTY_ZIP_UTIL_OBJS1) \
+	$(THIRD_PARTY_ZIP_UTIL_OBJS2)
+
+THIRD_PARTY_ZIP_COM_OBJS = \
+	o/$(MODE)/third_party/infozip/zip/zip.o 			\
+	o/$(MODE)/third_party/infozip/zip/zipfile.o 		\
+	o/$(MODE)/third_party/infozip/zip/zipup.o 			\
+	o/$(MODE)/third_party/infozip/zip/fileio.o 			\
+	o/$(MODE)/third_party/infozip/zip/util.o 			\
+	o/$(MODE)/third_party/infozip/zip/globals.o 		\
+	o/$(MODE)/third_party/infozip/zip/crypt.o 			\
+	o/$(MODE)/third_party/infozip/zip/ttyio.o 			\
+	o/$(MODE)/third_party/infozip/zip/unix/unix.o 		\
+	o/$(MODE)/third_party/infozip/zip/crc32.o 			\
+	o/$(MODE)/third_party/infozip/zip/zbz2err.o 		\
+	o/$(MODE)/third_party/infozip/zip/deflate.o 		\
+	o/$(MODE)/third_party/infozip/zip/trees.o 		
+
+THIRD_PARTY_ZIPSPLIT_OBJS = \
+	o/$(MODE)/third_party/infozip/zip/zipsplit.o 		\
+	$(THIRD_PARTY_ZIP_UTIL_OBJS1)
+
+THIRD_PARTY_ZIPNOTE_OBJS = \
+	o/$(MODE)/third_party/infozip/zip/zipnote.o 		\
+	$(THIRD_PARTY_ZIP_UTIL_OBJS1)
+
+THIRD_PARTY_ZIPCLOAK_OBJS = \
+	o/$(MODE)/third_party/infozip/zip/zipcloak.o 		\
+	o/$(MODE)/third_party/infozip/zip/ttyio.o 		\
+	$(THIRD_PARTY_ZIP_UTIL_OBJS1)					\
+	$(THIRD_PARTY_ZIP_UTIL_OBJS2)
+
+THIRD_PARTY_ZIP_LARGE_OBJS = \
+	o/$(MODE)/third_party/infozip/zip/zip.o			\
+	o/$(MODE)/third_party/infozip/zip/zipsplit.o	\
+	o/$(MODE)/third_party/infozip/zip/fileio.o		\
+	o/$(MODE)/third_party/infozip/zip/fileio_.o		
+
+o/$(MODE)/third_party/infozip/zip/%_.o: third_party/infozip/zip/%.c
```

Comment:
The `infozip.mk` recipe has the same file (for example `fileio.c`) compiled in two different ways: an extra `-DUTIL` flag is passed for `fileio_.o` but not for `fileio.o`. Does something like this happen in any other Makefile in the repo?

The workaround I used here was to create a special rule for the `%_.o` targets in this folder based on the default `%.o` rule.

Alternatively, I can just create copies of the source files (for example, `fileio_.c` as an exact copy of `fileio.c`) which would avoid the wildcard rule above, because I can just add `-DUTIL` to `OVERRIDE_CFLAGS` for the `_.o` files. @jart thoughts?

Reply:
It does happen. What you did here is fine. Another technique I normally use is this:

```make
$(THIRD_PARTY_MBEDTLS_A_OBJS):                                          \
                        OVERRIDE_CFLAGS +=                              \
                                -fdata-sections                         \
                                -ffunction-sections
```

Reply:
The problem I was running into was:

* `fileio.c` is part of the `zip` source code, `fileio_.c` is not
* `fileio.o` is a target that requires `fileio.c` to be compiled 
* `fileio_.o` is a target that requires the same `fileio.c` to be compiled with `-DUTIL` 
* the `%.o` rule in `build/rules.mk` expects `fileio_.c` to exist for the `fileio_.o` target, so adding `OVERRIDE_CFLAGS` alone does not work unless I create the missing `fileio_.c`

I think I have not properly understood how Makefile targets work; I shall lookup the docs and try some examples later.